### ISSUE TITLE
Look at compile times for generated code

### ISSUE THREAD

**Issue by thomasvl** [apple/swift-protobuf#63]
Since @allevato mentioned he'd seen some generated file crash the compiler and other take a while to compile…

Swift Weekly Issue 41 mentioned the [swift-dev thread](https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20161003/003099.html) on compile times.  It mentions some options (`-debug-time-function-bodies` & `-warn-long-function-bodies`), they might be useful to help see about improving the code generated.

**Comment by thomasvl** [comment]
More info/links:

  * http://irace.me/swift-profiling
  * http://mjtsai.com/blog/2016/10/06/reducing-swift-compile-time/
  * http://mjtsai.com/blog/2016/04/29/speeding-up-slow-swift-build-times/

Some comments about code changes that helped compile times in there and/or things linked off them.

**Comment by tbkka** [comment]
Ah, yes.  The `-debug-time-function-bodies` option and I are old friends.  ;-)  There's definitely still room for improvement, though.

**Comment by thomasvl** [comment]
@allevato I think you sorta did this recently.  I believe the answer was there wasn't anything really worth changing, but it might make sense to capture some of what you did find here.

I'm also not sure it is worth adding any of those compiler flags into the local builds, since there didn't see to be much we could do for things.

**Comment by allevato** [comment]
The biggest takeaways I found were that type checking was slowest for things involving global operators with a lot of overloads, and the ones we use most frequently in generated code are `~=` (indirectly, for matching `case`s) and `!=` (mostly in `isEqualTo`, but also to a lesser extent in the `has*` properties). So, large enums showed type checking slowness in their initializers (because we have a `case` for every raw value) and messages with large numbers of fields suffered in a handful of places.

There are horrible ways that I would *not* recommend to speed this up (basically create our own `notEqualTo` function to avoid the large global overload space for `!=`), but the improvements we're talking about would be on the order of milliseconds and it's not worth making our generated code horrible to get such a small gain.

We should see some improvements in Swift 4 for free. Thanks to [SE-0091](https://github.com/apple/swift-evolution/blob/master/proposals/0091-improving-operators-in-protocols.md) and [SE-0104](https://github.com/apple/swift-evolution/blob/master/proposals/0104-improved-integers.md), many of the current legacy global overloads of `==` and `!=` have been removed in favor of the generic one (https://github.com/apple/swift/commit/0b271e86d123ce5ec1190059cbb6a707363a542a), which will greatly reduce the load on the type checker.
